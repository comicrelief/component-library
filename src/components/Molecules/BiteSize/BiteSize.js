import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Link from '../../Atoms/Link/Link';
import { External, Internal } from '../../Atoms/Icons/index';
import spacing from '../../../theme/shared/spacing';
import zIndex from '../../../theme/shared/zIndex';
import hideVisually from '../../../theme/shared/hideVisually';
import Text from '../../Atoms/Text/Text';

const Container = styled.div`
  display: flex;
  position: relative;
  flex-direction: column;
  height: 100%;
  width: calc(100% - ${spacing('m')});
  color: ${({ theme }) => theme.color('grey_dark')};
  background: ${({ theme, backgroundColor }) => theme.color(backgroundColor)};
  @media ${({ theme }) => theme.breakpoint('small')} {
    flex-direction: row;
    width: 100%;
  }

  @media ${({ theme }) => theme.breakpoint('large')} {
    flex-direction: column;
  }
`;

const Heading = styled(Text)`
  text-transform: uppercase;
  letter-spacing: 0.03em;
`;

const Subtitle = styled(Heading)`
  margin: 0 0 ${spacing('m')};
`;

const Title = styled(Heading)`
  margin: 0 0 ${spacing('l')};
`;

const Copy = styled.div`
  padding: calc(2 * ${spacing('md')});
  ${({ hasLink }) => hasLink
    && `
    padding-bottom: calc(${spacing('m')} + ${spacing('l')});
  `};
  display: flex;
  flex-direction: column;
  border-radius: 1rem;
  box-shadow: 0px 0px 20px rgba(0, 0, 0, 0.15);
  background: ${({ theme, backgroundColor }) => theme.color(backgroundColor)};
  width: 100%;

  @media ${({ theme }) => theme.breakpoint('small')} {
    padding: calc(2 * ${spacing('m')});

    ${({ hasLink }) => hasLink
    && `
      padding-bottom: calc(${spacing('m')} + ${spacing('l')});
  `};
  }

  @media ${({ theme }) => theme.breakpoint('large')} {
    height: 100%;
  }
`;

const CTA = styled.div`
  position: absolute;
  right: ${spacing('m')};
  bottom: -${spacing('m')};
  ${zIndex('medium')};
`;

const Anchor = styled(Link)`
  height: 48px;
  max-width: 296px;

  ${({ isMobile }) => isMobile
    && `
      padding: ${spacing('sm')} calc(${spacing('md')} + ${spacing('l')});
      span {
        margin-left: 0;

        svg {
          width: 22px;
        }
      }
      span:first-child {
        hidden: ${hideVisually};
      }
    `}
`;

const BiteSize = ({
  backgroundColor,
  children,
  link,
  linkLabel,
  target,
  title,
  subtitle,
  ...rest
}) => {
  const [mobile, setMobile] = useState(false);

  const icon = () => {
    if (target === 'blank') {
      return <External colour="white" />;
    }
    return <Internal colour="white" />;
  };

  const external = target === 'blank' ? 'noopener noreferrer' : null;

  useEffect(() => {
    let width = typeof window !== 'undefined' ? window.innerWidth : null;
    function getWindowSize() {
      width = typeof window !== 'undefined' ? window.innerWidth : null;
      setMobile(width < 759);
    }

    if (typeof window !== 'undefined') {
      window.addEventListener('resize', getWindowSize);
    }
    setMobile(width < 759);
  }, []);

  return (
    <Container {...rest}>
      <Copy hasLink={link} backgroundColor={backgroundColor}>
        <Subtitle size="m" weight="normal" family="Anton" tag="h6">
          {subtitle}
        </Subtitle>
        <Title
          size="xl"
          tag="h4"
          height="2rem"
          weight="normal"
          uppercase
          family="Anton"
        >
          {title}
        </Title>
        {children}
      </Copy>
      {link && (
        <CTA>
          <Anchor
            isMobile={mobile}
            rel={external}
            color="red"
            href={link}
            target={target}
            type="button"
            icon={icon()}
          >
            <Text>{linkLabel}</Text>
          </Anchor>
        </CTA>
      )}
    </Container>
  );
};

BiteSize.propTypes = {
  backgroundColor: PropTypes.string,
  link: PropTypes.string,
  subtitle: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  linkLabel: PropTypes.string,
  target: PropTypes.string,
  children: PropTypes.node.isRequired
};

BiteSize.defaultProps = {
  backgroundColor: 'white',
  link: null,
  linkLabel: null,
  target: null
};

export default BiteSize;
