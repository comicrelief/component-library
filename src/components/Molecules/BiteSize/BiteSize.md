### BiteSize

```js
import Text from '../../Atoms/Text/Text';
import Link from '../../Atoms/Link/Link';
<div
  style={{
    display: 'flex',
    padding: '2rem 0',
    background: '#E1E2E3',
    justifyContent: 'space-around'
  }}
>
  <BiteSize
    target="blank"
    link="/home"
    linkLabel="find out more"
    backgroundColor="white"
    height="auto"
  >
    <Text tag="p">Text body copy description</Text>
  </BiteSize>
</div>;
```
