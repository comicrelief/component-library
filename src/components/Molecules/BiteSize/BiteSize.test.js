import React from 'react';
import 'jest-styled-components';
import renderWithTheme from '../../../hoc/shallowWithTheme';
import BiteSize from './BiteSize';

it('renders correctly', () => {
  const tree = renderWithTheme(
    <BiteSize
      target=""
      link="/home"
      linkLabel="Main action"
      backgroundColor="white"
      title="Heading"
      subtitle="Subtitle"
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
