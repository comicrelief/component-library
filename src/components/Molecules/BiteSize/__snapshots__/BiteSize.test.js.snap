// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c2 {
  font-size: 1.25rem;
  line-height: 1.25rem;
  text-transform: inherit;
  font-weight: normal;
  font-family: 'Anton',Impact,sans-serif;
  -webkit-letter-spacing: 0.03rem;
  -moz-letter-spacing: 0.03rem;
  -ms-letter-spacing: 0.03rem;
  letter-spacing: 0.03rem;
}

.c4 {
  font-size: 2rem;
  line-height: 2rem;
  text-transform: uppercase;
  font-weight: normal;
  line-height: 2rem;
  font-family: 'Anton',Impact,sans-serif;
  -webkit-letter-spacing: 0.03rem;
  -moz-letter-spacing: 0.03rem;
  -ms-letter-spacing: 0.03rem;
  letter-spacing: 0.03rem;
}

.c9 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-left: 1rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  right: 1.5rem;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  padding: 0.5rem 1.25rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  font-size: 1rem;
  border-radius: 2rem;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  height: 3.125rem;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background-color: #E52630;
  color: #FFFFFF;
}

.c7:hover {
  background-color: #890B11;
  color: #FFFFFF;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  width: calc(100% - 1.5rem);
  color: #222222;
  background: inherit;
}

.c3 {
  text-transform: uppercase;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  margin: 0 0 1.5rem;
}

.c5 {
  text-transform: uppercase;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  margin: 0 0 2rem;
}

.c1 {
  padding: calc(2 * 1rem);
  padding-bottom: calc(1.5rem + 2rem);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 1rem;
  box-shadow: 0px 0px 20px rgba(0,0,0,0.15);
  background: #FFFFFF;
  width: 100%;
}

.c6 {
  position: absolute;
  max-width: 296px;
  right: 1.5rem;
  bottom: -1.5rem;
  z-index: 2;
}

.c8 {
  height: 48px;
}

@media (min-width:740px) {
  .c10 {
    width: auto;
    right: auto;
    position: relative;
  }
}

@media (min-width:740px) {
  .c7 {
    width: auto;
  }
}

@media (min-width:740px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    width: 100%;
  }
}

@media (min-width:1440px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (min-width:740px) {
  .c1 {
    padding: calc(2 * 1.5rem);
    padding-bottom: calc(1.5rem + 2rem);
  }
}

@media (min-width:1440px) {
  .c1 {
    height: 100%;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <h6
      className="c2 BiteSize__Heading-sbjbnt-1 c3"
      color="inherit"
      size="m"
    >
      Subtitle
    </h6>
    <h4
      className="c4 BiteSize__Heading-sbjbnt-1 c5"
      color="inherit"
      height="2rem"
      size="xl"
    >
      Heading
    </h4>
  </div>
  <div
    className="c6"
  >
    <a
      className="c7 c8"
      color="red"
      href="/home"
      rel={null}
      target="_self"
      type="button"
    >
      <span
        className="c9"
        color="inherit"
        size="s"
      >
        Main action
      </span>
      <span
        className="c10"
        type="button"
      >
        <svg
          fill="#FFFFFF"
          heiht={24}
          viewBox="0 0 96 96"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M90.871 50.776c-.016.016-.026.036-.043.052l-32 32C58.047 83.609 57.024 84 56 84s-2.047-.391-2.829-1.171a4 4 0 010-5.657L78.343 52H8a4 4 0 010-8h70.343L53.171 18.829a4 4 0 015.657-5.657l32 32c.016.016.027.036.043.052.167.172.321.357.455.557.069.104.118.216.176.325.065.121.138.236.191.364.057.137.09.279.13.42.032.109.074.214.097.327a4.017 4.017 0 010 1.568c-.023.113-.065.217-.097.327-.041.141-.074.283-.13.42-.053.128-.126.243-.191.364-.059.109-.107.221-.176.325a4.035 4.035 0 01-.455.555z"
          />
        </svg>
      </span>
    </a>
  </div>
</div>
`;
